### ISSUE TITLE
Various fonts (including: Hurmit) glyphs offset/wrong size

### ISSUE THREAD

**Issue by Cpt0r** [ryanoasis/nerd-fonts#142]
#### Subject of the issue

Text and glyphs at 14pt:
<img width="123" alt="screen shot 2017-03-17 at 10 09 30" src="https://cloud.githubusercontent.com/assets/8690075/24027132/650684d0-0b19-11e7-9878-72ce33242039.png">

text at 14pt, glyphs at 18pt:
<img width="149" alt="screen shot 2017-03-17 at 10 10 28" src="https://cloud.githubusercontent.com/assets/8690075/24027134/6796d164-0b19-11e7-9061-fdcea0b2ef7e.png">

The glyphs seem to be either shifted on x and y axis, or just their size doesn't match text, or even both.

#### Your Setup

OSX, iTerm2, tmux.

**Comment by Cpt0r** [comment]
To clarify this issue isn't present in single-width version of the font.

**Comment by ryanoasis** [comment]
That's troubling 😟 but thanks for the report

**Comment by ryanoasis** [comment]
@sharkusk did a lot of great work fixing issues (in particular the mono issues) :smile: but I think we can carry over some of his logic to apply to _both_ mono and non-mono versions of fonts, in particular to:

1. vertical center the glyphs
2. stretch the powerline glyphs

I am pretty confident we want to do `#1` but not sure on doing `#2`

Here are some screenshots with my tweaks to the patcher:

![selection_013](https://cloud.githubusercontent.com/assets/8083459/24075602/1782a7c0-0bf5-11e7-851a-ca8505367e0f.png)

this is with `#1` change applied:
![selection_014](https://cloud.githubusercontent.com/assets/8083459/24075603/1b679fda-0bf5-11e7-8b6d-d5104033b923.png)

these are with `#1` and `#2` change applied:
![selection_015](https://cloud.githubusercontent.com/assets/8083459/24075604/1e70c2b0-0bf5-11e7-86c1-b93027924127.png)

![selection_016](https://cloud.githubusercontent.com/assets/8083459/24075624/6a33cb16-0bf5-11e7-8b52-50e924e51c08.png)

**Comment by Cpt0r** [comment]
More info:

The above was the `light` version.  Below is a shot of the `medium` which has slightly different offset:

<img width="176" alt="screen shot 2017-03-21 at 10 52 55" src="https://cloud.githubusercontent.com/assets/8690075/24126872/b453c0e4-0e24-11e7-8585-1537f81bdf24.png">

**Comment by ryanoasis** [comment]
@Cpt0r Thanks, I will test that version too

**Comment by ryanoasis** [comment]
This is working well so far but I have some more tweaking for the powerline glyphs to do.

Here is a before and after for Sauce Code Pro:

![selection_017](https://cloud.githubusercontent.com/assets/8083459/24225726/68475c46-0f38-11e7-9b99-0b44cb9f4a09.png)

**Comment by Cpt0r** [comment]
I noticed even in the fixed version there's more bottom edge space than top - is this intentional or are they not centered/using maximum available space?

I know nothing about fonts so this may very well be A Good Thing/The Right Way (tm).

**Comment by ryanoasis** [comment]
Not sure on that. I'll have to check in more detail. Thanks

**Comment by ddnexus** [comment]
Where could I try the font-patcher with the latest fixes/tries? I didn't find any branch/commit with that :(

**Comment by ryanoasis** [comment]
@ddnexus No branch yet. I'll try to create one this weekend or if not at least a gist diff

**Comment by ddnexus** [comment]
Thank you Ryan! A gist would be great too! Looking forward to try it soon. Thanks.

**Comment by ryanoasis** [comment]
Didn't get around to it today. Maybe tomorrow

**Comment by ddnexus** [comment]
Thank you Ryan.
Maybe a simple gist with the whole font-patcher file would be an easier option for you? Just a copy &paste, but that would allow us to patch the fonts right away. I cannot wait to fix how my terminal looks :)

**Comment by ryanoasis** [comment]
@ddnexus here you go: https://gist.github.com/ryanoasis/1b9a525aa3b5cad40c66d5e1d84c9b4d

very much a WIP (fixing a couple issues at the same time). If you manage to test it out let me know how it works for you...

I am considering also just getting a `dev`/`next` branch up that is up to date with WIP stuff... but not yet decided

**Comment by matthew-groth** [comment]
Thanks @ryanoasis, this has been bugging me for a long time. The current patch, however, does not fix it for me.

Before patch: 

![beforepatch](https://cloud.githubusercontent.com/assets/12831489/24783676/cfef431a-1b1b-11e7-83ff-9291a3d76061.png)

After Patch:
(Deleted font, used `git apply` on the gist you just uploaded, ran `./install.sh`, reloaded font)

![afterdiff](https://cloud.githubusercontent.com/assets/12831489/24783717/fd6268e0-1b1b-11e7-972b-9ecbe5e56706.png)

I don't notice any major difference.

My system:
OSX: 10.12.4
iTerm: Build 3.0.15
using Oh-My Fish with `agnoster` theme
Font: 24pt Liberation Mono Powerline Nerd Font Complete

**Comment by ddnexus** [comment]
Thank you Ryan,
this WIP is aleady a huge improvement in the look of the fonts. 

It looks like the Vertical and Horizontal centering is perfect for all glyphs, 
however the scaling still needs a few fixes: here is what I noticed till now, HTH you to continue with this great WIP:

- Powerline Extra Symbols
  - Vertical scale is OK
  - Horizontal scale is WRONG. The x scale ratio should be the same as y, instead
    the glyphs are stretched/shortened to the single char width.
    That is OK for the mono (single width variant) but not OK for the regular variant

- Seti-UI
  - The scale factor is huge! Especially compared to the Devicons.
    E.g. compare the dimensions of the python logo in the 2 sets and you will have a huge scale ratio difference.

- Devicons
  - The scale factor is small Especially compared to the Seti-UI
    E.g. compare the dimensions of the python logo in the 2 set and you will have a huge scale ratio difference.

While maximizing the scale to the full line height of the source font does make the glyphs as visible as possible,
it may make them look wrong when used with a line background as in a prompt. 

The glyphs that have the vertical dimension >= than the horizontal dimension will extend to the full line-height, hence
they will leave no space for the background above and below (graphically quite unpleasant).

The full line height is the correct size for the Powerline Extra Symbols (that have to play with the background), but for any other glyphs the scale factor should be calculated
with a small percentage less than the available line height. Maybe the calculation (of all but powerline Extra Symbols) should be done on the global collection of glyphs instead of group by group?
A good test would be e.g. the python logo in the Seti-UI and Devicon, which should result in quite the same dimensions.

Thank you for the great improvements till now! 

PS: A dev branch is certainly welcome, and may help eventual contributions. ;)

**Comment by ddnexus** [comment]
@matthew-groth: It doesn't look you patched the font using the font-patcher. Did you re-patch the font?

**Comment by matthew-groth** [comment]
Thanks @ddnexus you're right that I missed that step.

However, it still didn't work for me after running `./font-patcher "/Users/Matt/Library/Fonts/NerdFonts/Literation Mono Powerline Nerd Font Complete.ttf"`

[Here's the output from the font patcher.](https://github.com/ryanoasis/nerd-fonts/files/907072/FontPatcherOutput.txt)

There's still no difference between the font pre- and post-patch. After The patched font had "Nerd Font" appended to the name:

**Pre-Patch**
![image](https://cloud.githubusercontent.com/assets/12831489/24823418/23b7d024-1bcc-11e7-90b7-6f0c2e7b5c6b.png)
![image](https://cloud.githubusercontent.com/assets/12831489/24823432/4293b1a2-1bcc-11e7-902f-e1f8d11d77ee.png)

**Post-Patch**
![image](https://cloud.githubusercontent.com/assets/12831489/24823422/2c226a6c-1bcc-11e7-8bbe-89e61719da91.png)
![image](https://cloud.githubusercontent.com/assets/12831489/24823433/497f438c-1bcc-11e7-8c0f-309d691ee546.png)

By the way, the color changes are due to me changing my oh-my-fish theme to bobthefish.

Am I still missing a step?

**Comment by ddnexus** [comment]
It looks like you are patching an already patched font, missing a flag. You should patch the original not-yet-patched font, and use the -c flag.

**Comment by ryanoasis** [comment]
@matthew-groth Did you re-clear font cache? @ddnexus might be right (try patching an un-yet patched font). Also I would remove whatever nerd fonts you have, clear cache, run patcher on unpatched font, close terminals and re-open. Might this is overkill but might be something missing in those steps

---

Looking at your output all I can think is "really need to fix that progress bar" :blush:

**Comment by ryanoasis** [comment]
@ddnexus Thanks for the feedback on the WIP, I really appreciate it. I will try to reply to your comments sometime tomorrow. 

For now I am working on some test scripts to help make testing easier, preview:

![selection_031](https://cloud.githubusercontent.com/assets/8083459/24823621/75d0600e-1bce-11e7-970d-630082f32208.png)


Also I will push a `dev` branch up tomorrow

**Comment by matthew-groth** [comment]
Still no luck 😕

Here's what I did:

1. Deleted all my powerline fonts
2. cleared font cache with `sudo atsutil databases -remove`
3. ran `./font-patcher -c "/Users/Matt/Desktop/nerd-fonts/src/unpatched-fonts/LiberationMono/Literation Mono Powerline.ttf"`
4. copied `/Users/Matt/Desktop/nerd-fonts/Literation Mono Powerline Nerd Font Complete.ttf` to `~/Library/Fonts`
5. logged out and back in
6. opened up iTerm and set font to LiberationMonoPowerLine Nerd Font
7. restarted iTerm

![image](https://cloud.githubusercontent.com/assets/12831489/24824030/0b167866-1bd3-11e7-8265-e7a1bfcba932.png)


I must still be forgetting something?

**Comment by ddnexus** [comment]
It looks like you are using a Powerline patched font. Could you try to patch a normal font like Menlo?

**Comment by ddnexus** [comment]
And BTW... Literation or Liberation?

**Comment by ryanoasis** [comment]
@ddnexus Just fixed one of the points you raised:
> Horizontal scale is WRONG. The x scale ratio should be the same as y, instead the glyphs are stretched/shortened to the single char width. That is OK for the mono (single width variant) but not OK for the regular variant

You can see **mono (left side)** and **non-mono (right side)**:

![selection_032](https://cloud.githubusercontent.com/assets/8083459/24832794/c7186422-1c85-11e7-890d-5f740fe4c8b7.png)

**Comment by ddnexus** [comment]
That looks great! Thanks Ryan!

**Comment by ryanoasis** [comment]
@ddnexus Sure thank you. There are still some spacing problems as you can see but I think we are moving in a good direction. Also this is being updated in the [`dev`](https://github.com/ryanoasis/nerd-fonts/tree/dev) branch.

**Comment by Cpt0r** [comment]
@ryanoasis not sure what the proper etiquette is for closed issues but here's a quick update.

Using latest `dev` [8ab0affbee9c61c38a455b0d9c51bf2c01cc93b5] things are looking much better 👍.  Except for two things which keep grating my eyes: 

- the left facing powerline separator < is offset to the right, while the right facing > is perfect
- the git branch icon is offset to the right as well

<img width="78" alt="screen shot 2017-04-28 at 10 56 18" src="https://cloud.githubusercontent.com/assets/8690075/25510223/d3e8acda-2c01-11e7-9c3d-8423bea983c8.png">

<img width="150" alt="screen shot 2017-04-28 at 10 55 52" src="https://cloud.githubusercontent.com/assets/8690075/25510219/cfecd426-2c01-11e7-8c2e-f32022487570.png">

**Comment by ryanoasis** [comment]
@Cpt0r Thanks for the update 👍 . I'll see what I can do... it's getting there

By the way `master` has all those fixes and tweaks that dev had now.

Probably going to abandon dev for simpler workflow

**Comment by ryanoasis** [comment]
Actually yeah np on the etiquette. This was auto closed because of my commit message reference

**Comment by Cpt0r** [comment]
Hi @ryanoasis another update against the latest master 39db2fb0c255c75e27de59a4ed8f211a867be59e.

Hurmit font:

1. The powerline separators have regressed since last update:
<img width="643" alt="screen shot 2017-07-03 at 10 42 38" src="https://user-images.githubusercontent.com/8690075/27774812-a521bbd0-5fdc-11e7-88ca-4b1a831cb270.png">

2. I tried running the patcher myself against the Hermit and Input fonts but I get question mark boxes for all the icons - the Hurmit font in the repo is what I used above.
